### PR TITLE
Expand output tensor spec

### PR DESCRIPTION
**Pull request by FynnBe**
Implements changes discussed in https://github.com/bioimage-io/spec-bioimage-io/issues/252

### CODE REVIEW

**Comment by constantinpape** [comment]
@FynnBe this should work now. Please review and go ahead if you agree.

**Comment by FynnBe** [comment]
looks good. I have two follow-up questions that we could also address in a following PR, just making sure these aspects will not collide:
- [ ] how do we shrink a tensor with respect to its reference tensor?
- [ ] (should really be in another PR) we discussed in the past to allow also input tensors to be specified via another input tensor... this is not yet the case, is it??

File: bioimageio/spec/model/v0_4/schema.py [hunk 1]
```diff
@@ -456,23 +456,36 @@ def get_min_shape(t) -> numpy.ndarray:
...
-                tensors_by_name[out.shape.reference_tensor].shape
-            ):
-                raise ValidationError(
-                    f"Referenced tensor {out.shape.reference_tensor} "
-                    f"with {len(tensors_by_name[out.shape.reference_tensor].shape)} dimensions does not match "
-                    f"output tensor {out.name} with {len(out.shape)} dimensions."
-                )
+            if isinstance(out.shape, raw_nodes.ImplicitOutputShape):
+                ndim_ref = len(tensors_by_name[out.shape.reference_tensor].shape)
+                ndim_out = len([scale for scale in out.shape.scale if scale is not None])
```
Comment (FynnBe) [inline bioimageio/spec/model/v0_4/schema.py:482]: here `ndim_out` would include the `None` axis, but you compare to `ndim_ref`, so you leave it out.. maybe we can rename this here or comment on it?

File: bioimageio/spec/shared/schema.py [hunk 2]
```diff
@@ -113,7 +113,9 @@ def matching_lengths(self, data, **kwargs):
 class ImplicitOutputShape(SharedBioImageIOSchema):
```
Comment (FynnBe) [inline bioimageio/spec/shared/schema.py:113]: if we are very diligent we'd add validation such that offset is not zero where scale is `None`...

**Comment by FynnBe** [comment]
@constantinpape feel free to merge if you want to leave these for a future PR ;-)

**Comment by constantinpape** [comment]
> how do we shrink a tensor with respect to its reference tensor?

We haven't discussed this yet. As @oeway has correctly pointed out in https://github.com/bioimage-io/spec-bioimage-io/issues/252#issuecomment-936221835 we need a bit more work to cover all cases, which also include changing the axis order.
I suggest we skip the case of removing axes in the output for now and tackle it once we work on the more general case.

> (should really be in another PR) we discussed in the past to allow also input tensors to be specified via another input tensor... this is not yet the case, is it??

I think it's not implemented yet, but this one would be easy (just need to pass over all the tensors first and collect the names, instead of only going over the input tensors). And we need to add a few checks:
- no loop referencing of tensors (`input_a` <-> `input_b`)
- no referencing of inputs to outputs (`input` -> `output`)
- no referencing of outputs to outputs (`output_a` -> `output_b`)  (though we could also allow this one)

But it should be a separate PR.

**Comment by FynnBe** [comment]
> ... But it should be a separate PR.

yes, awesome! just came back here to say that the dropping of axes could just fall into the general mixed axes case, but I see you've been faster!

Let's merge once tests pass 👍

**Comment by FynnBe** [comment]
Instead of [df5ca66](https://github.com/bioimage-io/spec-bioimage-io/pull/442/commits/df5ca6650d441f5162e8c5fc618d62c223efa321) we could also update the expected error message in the test, but I thought only displaying it when relevant might avoid confusion...

**Pull request merged**